Sources/SourceControl/RepositoryManager.swift

```diff
@@ -165,54 +150,127 @@ public class RepositoryManager {
-            // Fetch and update the repository when it is being looked up.
-            // We need some mechanism to maybe not do this for pinned repos.
-            do {
-                let repo = try handle.open()
-                try repo.fetch()
-            } catch {
-                // FIXME: Better error handling.
-                print("Error while fetching the repo: \(handle)")
+    /// necessary.
+    public func lookup(repository: RepositorySpecifier, completion: @escaping (LookupResult) -> Void) {
+        DispatchQueue.global().async {
+            // Get the handle for this repository.
+            let handle = self.getHandle(repository: repository)
+            switch handle.status {
+
+            case .available:
+                let updatedResult = try! LookupResult {
+                    // Fetch and update the repository when it is being looked up.
+                    // We need some mechanism to maybe not do this for pinned repos.
+                    let repo = try handle.open()
+                    try repo.fetch()
+                    return handle
+                }
+                completion(updatedResult)
+
+            case .pending:
+                self.repositoriesLock.withLock {
+                    // Make sure we're still in pending status after aquiring this lock.
+                    guard handle.status == .pending else {
+                        // FIXME: This is wrong if handle is in error state.
+                        completion(Result(handle))
+                        return
+                    }
+                    // Just queue the completion handler in the handle if repository is being fetched right now.
+                    // They will be executed when repository is available.
+                    if let completionHandler = handle.completionHandler { 
+                        handle.completionHandler = { result in 
+                            completionHandler(result)
+                            completion(result) 
+                        } 
+                    } else { 
+                        handle.completionHandler = completion 
+                    }
+                }
+
+            case .uninitialized, .error:
+                // Fetch the repository.
+                let result = self.fetchRepository(of: handle)
+
+                // Save the manager state.
+                do {
+                    try self.saveState()
+                } catch {
+                    // FIXME: Handle failure gracefully, somehow.
+                    fatalError("unable to save manager state \(error)")
+                }
+
+                // Call the completion and any previous completion handlers we queued.
+                completion(result)
+                handle.completionHandler?(result)
+                // Release the handlers.
+                self.repositoriesLock.withLock {
+                    handle.completionHandler = nil
+                }
+            }
+        }
```

Comment:
I suspect this code can be simplified if you first always get the handle, and chain the completion handler, and then act on the state of the handle. That should eliminate the duplicate places the completion handler is called.